### ISSUE TITLE
[Question]: How to speed up loading&creating VectorStoreIndex?

### ISSUE THREAD

**Issue by TianyuFan0504** [run-llama/llama_index#12786]
### Question Validation

- [X] I have searched both the documentation and discord for an answer.

### Question

@dosu-bot

As the external data gets bigger and bigger the VectorStoreIndex seems to be getting very slow to create and read, I understand that it takes time to create but why is reading so slow as well? Is there any way to speed it up?

For example, I have a total of 1.5G of text files under '../datasets', which takes about 4000 seconds to create, and about 500s to save and read the created VectorStoreIndex.

Here is my code and settings
```
llm = OpenAI(temperature=0, 
             model="gpt-3.5-turbo"
             )
embed_model = HuggingFaceEmbedding(model_name='facebook/contriever')
Settings.llm = llm
Settings.embed_model = embed_model
Settings.node_parser = SentenceSplitter(chunk_size=2048, chunk_overlap=20)
Settings.context_window = 8192
Settings.num_output = 512
Settings.tokenizer = tiktoken.encoding_for_model("gpt-3.5-turbo").encode
```

```
documents = SimpleDirectoryReader("../datasets/").load_data()
nodes = Settings.node_parser.get_nodes_from_documents(documents)
#Creating:
index = VectorStoreIndex(nodes,include_embeddings=True)#it takes 4000s
index.storage_context.persist(persist_dir="../storage/")
#Reading:
index = load_index_from_storage(StorageContext.from_defaults(persist_dir="../storage/"))#it takes 500s
```

**Comment by TianyuFan0504** [comment]
@dosu-bot 
What about an acceleration method when reading VectorStoreIndex? Do you have any suggestions?

**Comment by TianyuFan0504** [comment]
@dosu-bot
The main time consumption seems to be in the part where getting nodes from documen (see the code), do you have any suggestions about speeding up this part?

```
documents = SimpleDirectoryReader("../datasets/").load_data()
nodes = Settings.node_parser.get_nodes_from_documents(documents)
```

**Comment by logan-markewich** [comment]
@TianyuFan0504 You are using a local model, so it is going to be limited by your hardaware. If you are running on CPU, it will be very slow (it should be running on GPU automatically if you have CUDA installed)

You can also try increasing the batch size for embeddings, but note you may run into memory issues if its too big

`embed_model = HuggingFaceEmbedding(model_name='facebook/contriever', embed_batch_size=20)`

**Comment by logan-markewich** [comment]
Oh, is the main time is in loading documents?

It really depends on what kinds of documents you are loading. You can try increasing the num workers

```
documents = SimpleDirectoryReader("../datasets/").load_data(num_workers=4)
```

**Comment by TianyuFan0504** [comment]
> Oh, is the main time is in loading documents?
> 
> It really depends on what kinds of documents you are loading. You can try increasing the num workers
> 
> ```
> documents = SimpleDirectoryReader("../datasets/").load_data(num_workers=4)
> ```

hi @logan-markewich, thanks for your help, which helps me lot : ) 
 
After further testing, I found that the function `get_node_from_documents` consumes almost 60% of the time, similar to what is mentioned in these 2 issues: https://github.com/run-llama/llama_index/issues/11025 and https://github.com/run-llama/llama_index/issues/11844.

If I understand correctly, it seems that this function doesn't support running in parallel? I'm trying to build a parallelizable version of this function, but I'm not sure if it's a duplicate of what's already out there.




```
documents = SimpleDirectoryReader("../datasets").load_data()
nodes = Settings.node_parser.get_nodes_from_documents(documents,show_progress = True)#consumes almost 60% of the time
index = VectorStoreIndex(nodes,include_embeddings=True)
```


BTW, I found that as the number of documents gets progressively larger, the time savings from specifying or not specifying  num_workers  in the function `SimpleDirectoryReader.load_data()` becomes less and less.  I'm looking into the cause of this.

**Comment by ksjadeja** [comment]
You are using simple SentenceSplitter and you can use pipeline(IngestionPipeline) to use the num_workers parameter and there would be significant speedup. The reason I am facing this issue is because of the way SemanticSplitterNodeParser is implemented and hence multiprocessing fails.

**Comment by TianyuFan0504** [comment]
It works! Thank you @ksjadeja : )
